src/main/java/net/dv8tion/jda/core/JDABuilder.java

```diff
@@ -104,11 +105,41 @@ public JDABuilder setReconnectQueue(SessionReconnectQueue queue)
     }
 
     /**
+     * Sets the {@link net.dv8tion.jda.core.ShardedRateLimiter ShardedRateLimiter} that will be used to keep
+     * track of rate limits across sessions.
+     * <br>When one shard hits the global rate limit all others will be informed by this value wrapper.
+     *
+     * <p>It is recommended to use the same ShardedRateLimiter for all shards and not one each. This is
+     * similar to {@link net.dv8tion.jda.core.requests.SessionReconnectQueue SessionReconnectQueue}!
+     *
+     * <p><b>This value is set when invoking {@link #setToken(String)} and cannot be unset using {@code null}. The re-use of this builder
+     * to build each shard is sufficient and setting it is not required.</b>
+     *
+     * <p>When you construct multiple JDABuilder instances to build shards it is recommended to use the same ShardedRateLimiter on
+     * all of them. But it is to be <u>avoided</u> to use the same ShardedRateLimiter for different accounts/tokens!
+     *
+     * @param  rateLimiter
+     *         ShardedRateLimiter used to keep track of cross-session rate limits
+     *
+     * @return The {@link net.dv8tion.jda.core.JDABuilder JDABuilder} instance. Useful for chaining.
+     */
+    public JDABuilder setShardedRateLimiter(ShardedRateLimiter rateLimiter)
+    {
+        if (accountType != AccountType.BOT)
+            this.shardRateLimiter = null;
+        else
+            this.shardRateLimiter = rateLimiter == null ? new ShardedRateLimiter() : rateLimiter;
+        return this;
+    }
+
+    /**
      * Sets the token that will be used by the {@link net.dv8tion.jda.core.JDA} instance to log in when
      * {@link net.dv8tion.jda.core.JDABuilder#buildAsync() buildAsync()}
      * or {@link net.dv8tion.jda.core.JDABuilder#buildBlocking() buildBlocking()}
      * is called.
      *
+     * <p><u><b>This will reset the prior provided {@link #setShardedRateLimiter(ShardedRateLimiter)} setting!</b></u>
+     *
```

Comment:
Whoa! This is potentially dangerous.
I suggest a `shardedRateLimiterSet` boolean to account for user code calling that first, then `setToken`, which resets the rate limiter.
Or is this for if you reuse the builder for multiple tokens? 🤔😲

Anyway, if all else fails I would at least detect if users call `setShardedRateLimiter` and then `setToken` and log that, to notify that we're setting up a new rate limiter.

(setters with side effects are tricky, because principle-of-least-astonishment)

Reply:
This isn't dangerous, there is a reason this warning is bold-underlined in the documentation. Most users will not even use this setter in the first place.

Reply:
I'm just saying, the few people who will use it *better* not do it in the wrong order.
Since it will ignore the rate limiter that they've set.